### ISSUE TITLE
Does rbox work for all SVG elements ?

### ISSUE THREAD

**Issue by louisameline** [wout/svg.js#397]
Hi and thank you for the library.

I've never used SVG but I maintain a tooltip library and some users wish they could put tooltips on SVG elements. I noticed it was hard to find the actual coordinates of an SVG path on screen, and found your library during my research.

My question is in the title: does rbox work for all SVG elements (circle, polygons, etc) or only for paths? I'm trying to decide if it's reasonable to include the calculations of SVG (size and position) in our library, or leave it altogether to external libraries like SVG.js.

Thank you

**Comment by Fuzzyma** [comment]
Yes the `rbox()` method is available on all elements.
However I just did a few tests and it turns out that the viewbox method has a bug and therefore the rbox goes infinite. You should be fine if you are not using nested viewboxes.

If size and position is the only thing you wanna do with the whole svgjs-lib its probably better to create your own piece of Code which calculates your data. Actually the Browser is doing all the work for you:

```javascript
var rectBox = svgElement.getBoundingClientRect()

// will display a div at exactly the right spot. Note that I use `fixed` because
// getBoundingClientRect depends on scroll
$('<div>').css({
  position:'fixed',
  width:rectBox.width,
  height:rectBox.height,
  top:rectBox.top,
  left:rectBox.left,
  outline:'3px solid black'
}).appendTo('body')
```

**Comment by louisameline** [comment]
Thank you for the response. I already use getBoundingClientRect but I believe it does not account for transformations like scaling, moving, rotating, etc. I'm open to other ideas because I'm struggling a little to extract the necessary code and I did not find a stand-alone script for that purpose.

Can you please confirm something else for me: if the SVG element has been created with Raphael.js or something else, there is no way to use svg.js's rbox on it, is there ? Thank you.

**Comment by Fuzzyma** [comment]
`getBoundingClientRect` **does** include any transformation and also viewbox.

Just try it:

```
var draw = SVG('your_id')

// some viewbox stuff
var inner = draw.nested().viewbox(0,0,300, 200).nested().viewbox(-10,-20,500, 500)

// scaled, rotated
var rect = inner.rect(100,100).move(100,100).fill('none').stroke('#000').rotate(45).scale(2)
var rectBox = rect.node.getBoundingClientRect()

$('<div>').css({
  position:'fixed',
  width:rectBox.width,
  height:rectBox.height,
  top:rectBox.top,
  left:rectBox.left,
  outline:'3px solid black'

}).appendTo('body')
```

For the other thing there is a solution:

```
var myEl = SVG.get('id_of_svg').select('css_selector as you which').first()

// e.g.
var polygon = SVG.get('drawing').select('polygon').first()
// or
var polygon = SVG.get('drawing').select('#id_of_polygon').first()
```

**Comment by louisameline** [comment]
Your example with the `rect` works fine but I can't get the right results with a `path`.
In this jsFiddle (provided by one of my users), I tried to frame the purple path, using `rbox` with the technique you gave me (which actually gives the same results as `getBoundingClientRect`) : https://jsfiddle.net/Lq0uvL74/7/

Chrome is kinda wrong and Firefox is completely lost. Am I doing something wrong?

**Comment by Fuzzyma** [comment]
I tried a path and it seems like it works to, **but** the box of the path is rotated and the box of this rotated box is the clientRect of the path and not the path itself. That leads to confusion because its not what one want (just imagine you would rotate a rect - the resulting bbox is bigger than the orginal). The center of this both shapes should be close to each other so maybe you can use it anway.

However I found a way to get the real real real bbox in screen coordinates from an element. Unfortunately this method requires to transform all points of the shape into the screen coordinate system with the help of `getScreenCtm()`, followed by drawing this points and getting the bbox of this newly drawed shape.
Thats need a lots of processing time and is stupid to implement because you have to catch every case of path attributes. Maybe I will implement that in another plugin when I find the time but for now just an example with a rly simple path which doenst need all this pitfall checks:

```javascript
var inner = draw.nested().viewbox(0,0,300, 200).nested().viewbox(-10,-20,500, 500)

var rect = inner.path('M100,100 200,200 500,100 200,400 100,100').move(100,100).fill('none').stroke('#000').rotate(45).scale(0.5)

var pt  = draw.node.createSVGPoint();
var m = rect.screenCTM().native()
var pathArr = rect.array().valueOf()

for(var i in pathArr){

  pt.x = pathArr[i][1]
  pt.y = pathArr[i][2]
  var temp = pt.matrixTransform(m)
  pathArr[i][1] = temp.x
  pathArr[i][2] = temp.y

}

var bbox = (new SVG.PathArray(pathArr)).bbox()

$('<div>').css({
  position:'fixed',
  width:bbox.width,
  height:bbox.height,
  top:bbox.y,
  left:bbox.x,
  outline:'3px solid black'

}).appendTo('body')

```

**Comment by louisameline** [comment]
When I try to apply this on my Raphael example, I get an error that says that rect.screenCTM is not defined, see https://jsfiddle.net/Lq0uvL74/10/. But that's ok, anyway I have my answer: no, I will not try to include the fix in our library, this is too complex for us to support. I'll just wait for your plugin if you have time and/or leave people do the maths themselves.

And just for the record, I tried Raphael's `Raphael.pathBBox(path)` method ; I got different results but not the ones I was hoping for either. Thank you

**Comment by Fuzzyma** [comment]
screenCTM was introduced after 2.1.1. I just released 2.2.0 with which the example should work.

**Comment by Fuzzyma** [comment]
I think that was fast enough: https://github.com/Fuzzyma/svg.screenbbox.js ;)
I only added path, polyline and polygon to the code because `getBoundingClientRect` seems to work with basic shapes (?)

Just use it as described in the readme wih `var bbox = path.screenBBox()`

**Comment by louisameline** [comment]
That was fast yes, thank you :)

The values returned by screenBBox seem only accurate the first time you ask them though, there are probably some values that you should reset at each recalculation.

Another issue I'm facing is that when I try to use it on a path created by Raphael.js, I get an error saying that `SVG.parser` is undefined.

My last issue is the difficulty to select an SVG element with `SVG.get('id_of_svg').select('css_selector as you which').first()`. My tooltip library has no garanty that the `<svg>` or `<path>` tags have ids or classes to select them, and I'd rather not add ids myself. My library just gets called when a mouseover (or click, etc) event is triggered on an element which is supposed to display a tooltip, so ideally I'd like to be able to just provide the node of the SVG element, I don't know if it's possible:

```
$(pathNode).on('mouseover', function(){
    var path = SVG.select(this),
        bbox = path.screenBBox();
        ...
});
```

Thanks !

**Comment by Fuzzyma** [comment]
I dont save any data on the node/instance. The calculation is the same every time. Maybe its an reference-issue when pulling the array from the path.
Test this for me:
Change the line `var arr = this.array().valueOf()` to `var arr = this.array().valueOf().slice()` and see if the error goes away.

The issue with the parser is harder to tame. When SVG creates a new instance it also creates a second instance what we call the parser. We use it to calculate the bbox of some shapes when they are not visible (e.g. in the defs - throws error in FF). The patharray-bbox method also uses this parser. Try to call `SVG.prepare()` before using the parser. 

The last thing should go away when you use `SVG.adopt(elementNode)`. It returns an svg instance of the given element. Try `SVG.adopt(svgNode).select('path').first()` for a result. I actually dont know if `SVG.adopt(this)` would also work. Just try it :)

**Comment by louisameline** [comment]
Right, but slice keeps references when items are objects so it's not working here but `var arr = JSON.parse(JSON.stringify(this.array().valueOf()))` solves the issue (but JSON is not supported in IE7 or lower...).

`SVG.prepare()` fixes the second issue indeed.

`SVG.adopt(this).screenBBox()` works, thank you, that's perfect. `SVG.adopt(this).select()` actually does not work because the object returned by `adopt()` has no `select` property.

The positioning is now better, expecially in Firefox where `getBoundingClientRect` was dramatically wrong. Last issue though, I assume that the width of the stroke is not taken into account during the calculations. In the jsFiddle I initially provided, the thickness of the circle creates a small offset for example.

**Comment by Fuzzyma** [comment]
Are you using paths for the circle (arc especially)? If so: It can happen, that the transformed bbox of an arc is wrong because I only transform the start and the end point of the circle. The fact, that the whole circle is skewed or something like this gets lost. So it could be better to use cubic bezieres instead.
However as far as i know bbox never takes stroke into account. But the center should be still the center.
`select()` only works on parent elements (svg, groups). 

I dont like the JSON solution. So much computing for nothing. Maybe I should just map slice like this:

```javascript
var array = something.map(function(el){
  return el.slice()
})
```

**Comment by louisameline** [comment]
Ok, we'll do with the small imperfections then. I don't like the JSON thing either, it was just the quickest fix for me. `map` is ecmascript 5 and does not work in IE8 and lower. But you'll easily find a deep copy function out there.

Thanks again for all the great work you did. I'll include the necessary code in my library so that tooltips work well out of the box if I detect that svgjs and screenBBox are available, and tell my users about it.

**Comment by Fuzzyma** [comment]
You're welcome :)
I fixed the last things and pushed v0.1.2 which will be the actual version for a while now I guess :D
